flink-yarn/src/main/java/org/apache/flink/yarn/YarnApplicationFileUploader.java

```diff
@@ -360,11 +362,27 @@ List<String> registerProvidedLocalResources() {
                     envShipResourceList.add(descriptor);
 
                     if (!isFlinkDistJar(filePath.getName()) && !isPlugin(filePath)) {
-                        classPaths.add(fileName);
+                        URI parentDirectoryUri = new Path(fileName).getParent().toUri();
+                        String relativeParentDirectory =
+                                new Path(filePath.getName())
+                                        .toUri()
+                                        .relativize(parentDirectoryUri)
+                                        .toString();
+
+                        if (!resourcesDir.contains(relativeParentDirectory)) {
+                            resourcesDir.add(relativeParentDirectory);
+                        }
+                        resources.add(fileName);
```

Comment:
I am wondering if naively adding all the files that are not Flink dist-jar or plugin as a resource file is over killing. One potential solution more accurate / explicit is to do something similar to the plugins. i.e. introduce a reserved keyword of "resources" as a dir name. All the files in this dir will be treated as resources.

BTW, given that this patch introduces a user sensible behavior change, we need a FLIP.

Reply:
The existing config `yarn.provided.lib.dirs` supports adding all the resources (jars or configs) under each of those dirs separated by comma in to the classpath. Why adding the parent dir also to the classpath is an overkill?

Take `yarn.ship-files` as an example takes arbitrary list of dirs and adds both the files and the dir under each of the individual dirs to the classpath. Isn't this the case with `yarn.provided.lib.dirs` as well? Except that in the `yarn.provided.lib.dirs` is more of a platform specific libs which won't change between different flink apps.

Also introducing a new `resources` dir and only supporting that feels limiting. Ideally, users or platforms would like to logically group resources in to their own dirs. For eg: hadoop, hive libs etc in to their own dirs.

Reply:
The main concern is the length of class path. I think there is a limit of 32K characters. For one of my simple local job, the class path is already over 10K chars. If we are including all the parent dir into the class path. This might easily go beyond the limit. Also, in general, adding the resource dir to class path in a more explicit and accurate way should probably be preferred.

> Also introducing a new resources dir and only supporting that feels limiting. Ideally, users or platforms would like to logically group resources in to their own dirs. For eg: hadoop, hive libs etc in to their own dirs.

Maybe we can have a `yarn.provided.resources.dirs` in this case.

Reply:
One thing to add here is, we are already `relativizing` the path so there are no redundant top level directory names added to the classpath. For eg: `/users/test/lib/hadoop/;/users/test/lib/hive` - only `.:hadoop:hive:` will be added to the classpath.

Note: the same applies to existing `yarn.ship-files` config as well which is working fine without issues.

> Also, in general, adding the resource dir to class path in a more explicit and accurate way should probably be preferred.
Do you mean resources as anything other than `jar` like `.xml`, `.yaml` etc should be added through `yarn.provided.resources.dirs`?

2 things:

1. Given this separation, won't the same issue still happen for `*.jar`?
2. I'm not super sure about introducing an additional config for this case.